### ISSUE TITLE
How font priority is decided when same version exist in userspace and system?

### ISSUE THREAD

**Issue by abhidesh128** [FontManager/font-manager#242]
Let's say, I have a font installed in userspace, i.e. in my `~/.local/share/fonts` or `~/.fonts`.
Now, I installed a font package that contains the same font and same version.
In this case, which font file will be actually preferred by `font-manager` or system apps? I experimented about it by right-clicking a font, and viewing it in its folder. Unless a font only installed in userspace, it always showed system's `/usr` folder. So system version of that font file is preferred by apps?

I know, this may not be specific to `font-manager` itself, but I guess the developers may know about this behaviour and can guide me about it.

**Comment by JerryCasiano** [comment]
This is determined by Fontconfig.

I have not delved into that area of their source code so I suggest you ask on their mailing list as that is most likely to get you a quick and correct response.

https://lists.freedesktop.org/mailman/listinfo/fontconfig

If I had to guess, I would expect it ignores duplicates and whichever one happened to be cached first is the one used, but again that would be a guess. Ask on their list.

**Comment by abhidesh128** [comment]
Thanks @JerryCasiano for your response..!

**Comment by abhidesh128** [comment]
Hii @JerryCasiano...! Before asking about this on Fontconfig mailing list, I decided to perform some experiment myself considering your guess. And yes. You were right. What I found is :

1. Fontconfig's matching mechanism is indeed complex. Complex nature of font data types and metrics even make it more confusing.
2. I read fontconfig's manpages and documentations. During this, somewhere I read that before performing any match, fontconfig sources the config files from system and user directories. In the main `fonts.conf` file in `/etc`, we can define the font directories where we keep `.ttf` or `.otf` files. I found that in this list, system directories are listed first, and then user directories are listed. Thus, it seems that fontconfig always refers to system directories first, and then user ones, sequencing same in the caches as well.
3. If a font file is in `/usr`, and the same file is in userspace, then the system one is preferred first. If we modify the file name in the user directory, then it magically appears first and overrides system's file. Thus, we can certainly say that the files's `mtime` is considered during caching, and *it is listed in the order it appeared in cache.*
4. Moreover, fontconfig gives unique IDs to every font directory and it's subdirectories. Any change in the parent directory is noted first. So the file whose name we modified must reside in the parent directory itself. Then only it will appear first in cache. So directory structure also matter in priority..! For ex. a file in `~./fonts` will be preferred first instead of a file in `~./fonts/foo`
